README.md

```diff
@@ -1,36 +1,83 @@
+bundling into "batches" of items like xargs does to amortize child set-up times.
 
 ## Motivating example
 
-Let's say you want to run formatting steps, like `isort` then `black` on some
-sources, and importantly, *in that order*.  If you had 100 source files and 2
-cores, you could probably saturate your ability to run in in parallel by
-running all the isorts first, then all the blacks.
+First, let's say we're operating on keys that are filenames, and values that
+are the bytes in those files.  We want to perform a series of steps in order,
+like running various autofixing lint engines that might produce conflicting
+diffs if run in parallel.  So we run them sequentially for a given key.
+
+```py
+def func(k, v):
+    return (k, engine3(engine2(engine1(v))))
+
+stream = [...]
+with ThreadPoolExecutor() as t:
+    result = t.starmap(func, stream)
+```
+
+Although this is nice and predictable, there are two major downsides:
+
+1. They're nested, so the total time is _at least_ the sum of all engines'
+   times for a given key even if you had a million cores.  There's typically a
+   large variance in runtimes, and this stacks them in the worst way (a big
+   file is going to be slow in all engines).
+2. If we get an exception, we basically lose that key entirely.  A much better
+   behavior would be that if `engine2` raised and exception on something, we just
+   skip `engine2` for all keys (including ones that already have `engine3` run on
+   them).
+
+In feedforward, you just need minimal wrapping and to let it rip:
+
+```py
+class MyStep(feedforward.Step):
+    def __init__(self, func):
+        self.func = func
+
+    def match(self, key):
+        return True
+
+    def process(self, gen, notifications):
+        for n in notifications:
+            self.output_queue.append(self.update_notification(n, new_value=self.func(n.value)))
+
+r = Run()
+r.add_step(MyStep(engine1))
+r.add_step(MyStep(engine2))
+r.add_step(MyStep(engine3))
+results = r.run_to_completion(files)
+```
+
+## Data Types
 
-But what if there's 1 source file or 100 cores?  If you're going to have a spare
-core just sitting around, it may as well run `black` on the unmodified version,
-and only keep the result if `isort` ended up not making changes.
+Keys must be immutable and hashable, as they are commonly used in dict keys.
 
-We can improve the best-case runtime on a given file from `A+B` to `max(A, B)`
-assuming sufficient cores and B being scheduled early enough to fully hide its
-latency, and the worst-case runtime remains approximately `A+B`.
-Interestingly, the worst-case holds regardless of whether you're saturating all
-your cores, because as you saturate you also mispredict less.
+Values are less restricted, but using more words:
 
-The running in parallel is the easy part; cache invalidation is the hard part.
+* They need to be equality-comparable (but false-negatives are OK)
+* They need have a consistent repr
```

Comment:
I'm not sure what an inconsistent repr means, or why we require something of the repr?